File: packages/core/src/land/units.ts
/**
 * Land-area units (ported from the predecessor's pattadar app's units.ts concepts).
 * Canonical unit is the ACRE: `parcel.extent` is decimal acres and `unit` is
 * only a provenance label. Pure functions, no dependencies.
 */

export type UnitKey =
  | 'acre'
  | 'cent'
  | 'gunta'
  | 'sqyd'
  | 'sqft'
  | 'sqm'
  | 'hectare'
  | 'ankanam';

/** Square feet per 1 unit (unambiguous base). */
export const UNIT_SQFT: Record<UnitKey, number> = {
  sqft: 1,
  sqyd: 9,
  sqm: 10.7639104,
  ankanam: 72, // AP/Telangana common value
  cent: 435.6, // 43560 / 100
  gunta: 1089, // 43560 / 40 (= 2.5 cents)
  acre: 43560,
  hectare: 107639.104,
};

const SQFT_PER_ACRE = 43560;

/** Display order for pickers / converter grid. */
export const UNITS: { key: UnitKey; label: string; short: string }[] = [
  { key: 'acre', label: 'Acres', short: 'ac' },
  { key: 'cent', label: 'Cents', short: 'ct' },
  { key: 'gunta', label: 'Guntas', short: 'g' },
  { key: 'sqyd', label: 'Sq. yards', short: 'sq.yd' },
  { key: 'sqft', label: 'Sq. feet', short: 'sq.ft' },
  { key: 'sqm', label: 'Sq. metres', short: 'sq.m' },
  { key: 'hectare', label: 'Hectares', short: 'ha' },
  { key: 'ankanam', label: 'Ankanam', short: 'ank' },
];

export const round2 = (n: number): number =>
  Math.round(((Number(n) || 0) + Number.EPSILON) * 100) / 100;

export function toAcres(value: number, unit: UnitKey): number {
  return ((Number(value) || 0) * UNIT_SQFT[unit]) / SQFT_PER_ACRE;
}

export function fromAcres(acres: number, unit: UnitKey): number {
  return ((Number(acres) || 0) * SQFT_PER_ACRE) / UNIT_SQFT[unit];
}

export function convert(value: number, from: UnitKey, to: UnitKey): number {
  return fromAcres(toAcres(value, from), to);
}

export function acresToAll(acres: number): { key: UnitKey; label: string; value: number }[] {
  return UNITS.map((u) => ({ key: u.key, label: u.label, value: fromAcres(acres, u.key) }));
}

/** "2 Acres 50 Cents" — cents keep a decimal only when non-integer. */
export function formatArea(acres: number): string {
  const a0 = Number(acres) || 0;
  if (a0 <= 0) return '0 Cents';
  const totalCents = round2(a0 * 100);
  const ac = Math.floor(totalCents / 100 + 1e-9);
  const ct = round2(totalCents - ac * 100);
  const parts: string[] = [];
  if (ac > 0) parts.push(`${ac} ${ac === 1 ? 'Acre' : 'Acres'}`);
  if (ct > 0 || ac === 0) parts.push(`${round2(ct)} ${ct === 1 ? 'Cent' : 'Cents'}`);
  return parts.join(' ');
}

/** "4 Acres 34 Guntas" — the guntas view of an acre value (40 guntas/acre). */
export function formatAcresGuntas(acres: number): string {
  const a0 = Number(acres) || 0;
  if (a0 <= 0) return '0 Guntas';
  const totalGuntas = round2(a0 * 40);
  const ac = Math.floor(totalGuntas / 40 + 1e-9);
  const g = round2(totalGuntas - ac * 40);
  const parts: string[] = [];
  if (ac > 0) parts.push(`${ac} ${ac === 1 ? 'Acre' : 'Acres'}`);
  if (g > 0 || ac === 0) parts.push(`${round2(g)} ${g === 1 ? 'Gunta' : 'Guntas'}`);
  return parts.join(' ');
}

/** Map a stored unit label (legacy compound or key) to a UnitKey. */
export function unitKey(label: string | null | undefined): UnitKey {
  const s = String(label || '')
    .toLowerCase()
    .trim();
  if ((s as UnitKey) in UNIT_SQFT) return s as UnitKey;
  if (s.includes('acre')) return 'acre'; // legacy "Acres-Guntas" is decimal acres
  if (s.includes('gunta')) return 'gunta';
  if (s.includes('cent')) return 'cent';
  if (s.includes('hect')) return 'hectare';
  if (s.includes('ankan')) return 'ankanam';
  if (s.includes('yard') || s.includes('sq.yd') || s.includes('sqyd')) return 'sqyd';
  if (s.includes('metre') || s.includes('meter') || s.includes('sq.m') || s.includes('sqm')) return 'sqm';
  if (s.includes('feet') || s.includes('sq.ft') || s.includes('sqft')) return 'sqft';
  return 'acre';
}

export function unitLabel(label: string): string {
  const u = UNITS.find((x) => x.key === unitKey(label));
  return u ? u.label : label;
}

/** Global extent-display preference (CL-13). */
export type ExtentPref = 'acres-cents' | 'acres-guntas' | 'cents' | 'sqyd';

/** THE extent formatter (CL-14) — every surface renders through this. */
export function formatExtent(acres: number, pref: ExtentPref = 'acres-cents'): string {
  const a = Number(acres) || 0;
  switch (pref) {
    case 'acres-guntas':
      return formatAcresGuntas(a);
    case 'cents':
      return `${round2(a * 100).toLocaleString('en-IN')} Cents`;
    case 'sqyd':
      return `${Math.round(a * 4840).toLocaleString('en-IN')} Sq.yd`;
    default:
      return formatArea(a);
  }
}

/** Natural sort for survey numbers: numeric-aware, treats "/" and "-" alike
 * ("1/2" < "1/10"; "126/2" == "126-2" positionally). */
export function naturalCompare(a: string, b: string): number {
  const split = (s: string) => (s || '').split(/[\/\-.]/).map((p) => p.trim());
  const pa = split(a);
  const pb = split(b);
  for (let i = 0; i < Math.max(pa.length, pb.length); i++) {
    const x = pa[i] ?? '';
    const y = pb[i] ?? '';
    if (x === y) continue;
    const nx = Number(x);
    const ny = Number(y);
    if (Number.isFinite(nx) && Number.isFinite(ny)) return nx - ny;
    return x.localeCompare(y, undefined, { numeric: true });
  }
  return 0;
}

File: apps/ios/PattadarKit/Sources/PattadarKit/Land/Units.swift
import Foundation

/// Land-area units. The canonical unit is the ACRE: a parcel's `extent` is
/// decimal acres and its `unit` is only a provenance label saying how the
/// number was written down.
///
/// This mirrors `packages/core/src/land/units.ts` exactly, and is held to it by
/// `UnitVectorTests` reading `vectors/units.json`. That is not ceremony: the
/// two implementations have already drifted once inside a single language —
/// one screen converted through `toAcres` and another wrote the raw number, so
/// a scanned "40 Guntas" would have been filed as forty ACRES. Across a
/// language boundary the same mistake is invisible until someone's holding is
/// overstated by a factor of forty.
public enum UnitKey: String, CaseIterable, Sendable {
    case acre, cent, gunta, sqyd, sqft, sqm, hectare, ankanam

    /// Square feet in one of this unit — the unambiguous base everything
    /// converts through.
    public var squareFeet: Double {
        switch self {
        case .sqft: 1
        case .sqyd: 9
        case .sqm: 10.7639104
        case .ankanam: 72          // the AP/Telangana value
        case .cent: 435.6          // 43560 / 100
        case .gunta: 1089          // 43560 / 40
        case .acre: 43560
        case .hectare: 107639.104
        }
    }

    public var label: String {
        switch self {
        case .acre: "Acres"
        case .cent: "Cents"
        case .gunta: "Guntas"
        case .sqyd: "Sq. yards"
        case .sqft: "Sq. feet"
        case .sqm: "Sq. metres"
        case .hectare: "Hectares"
        case .ankanam: "Ankanam"
        }
    }
}

private let squareFeetPerAcre = 43560.0

/// Resolve a free-text unit label — as a passbook, a deed or the AI extractor
/// wrote it — to a known unit.
///
/// Order matters and is load-bearing: the legacy label "Acres-Guntas" contains
/// both words, and it means DECIMAL ACRES, so acre must be tested first. Every
/// parcel currently in the database carries that label; reading it as guntas
/// would divide every holding by forty.
public func unitKey(_ label: String?) -> UnitKey {
    let s = (label ?? "").lowercased().trimmingCharacters(in: .whitespacesAndNewlines)
    if let exact = UnitKey(rawValue: s) { return exact }
    if s.contains("acre") { return .acre }
    if s.contains("gunta") { return .gunta }
    if s.contains("cent") { return .cent }
    if s.contains("hect") { return .hectare }
    if s.contains("ankan") { return .ankanam }
    if s.contains("yard") || s.contains("sq.yd") || s.contains("sqyd") { return .sqyd }
    if s.contains("metre") || s.contains("meter") || s.contains("sq.m") || s.contains("sqm") { return .sqm }
    if s.contains("feet") || s.contains("sq.ft") || s.contains("sqft") { return .sqft }
    // An unrecognised label is treated as acres because that is what the
    // canonical column already holds; guessing a smaller unit would silently
    // shrink the holding.
    return .acre
}

public func toAcres(_ value: Double, _ unit: UnitKey) -> Double {
    value * unit.squareFeet / squareFeetPerAcre
}

public func fromAcres(_ acres: Double, _ unit: UnitKey) -> Double {
    acres * squareFeetPerAcre / unit.squareFeet
}

public func convert(_ value: Double, from: UnitKey, to: UnitKey) -> Double {
    fromAcres(toAcres(value, from), to)
}

// MARK: - The two entities do NOT store area the same way

/// A parcel's `extent` is canonical DECIMAL ACRES and its `unit` is only a
/// provenance label. A property's `landArea` is the number as WRITTEN, in
/// `landUnit` — 418.5 with unit "Sq.yd" means 418.5 square yards, not 418.5
/// acres.
///
/// Getting this backwards rendered a 418 sq.yd plot as "2,025,420 Sq.yd" on
/// the dashboard, and silently inflated every acre subtotal that mixed the two.
/// These two helpers exist so no screen has to remember which convention
/// applies to which table.

/// Acres for a property, for totalling and sorting alongside parcels.
public func propertyAcres(landArea: Double, landUnit: String) -> Double {
    toAcres(landArea, unitKey(landUnit))
}

/// A property's area as it should be READ — in the unit it was filed in.
/// Converting a 200 sq.yd site to "0.04 ac" is correct and useless.
public func propertyAreaText(landArea: Double, landUnit: String) -> String {
    guard landArea > 0 else { return "" }
    let u = unitKey(landUnit)
    return String(format: "%g %@", landArea, u.label)
}

// MARK: - Each kind of holding has its own unit

/// What a holding is measured in, in India.
///
/// Farmland is acres. An open plot or site is square yards. A house, flat or
/// shop is square FEET of built-up area — "2,020 sq ft at Rajapushpa" is how
/// that is said, and nobody quotes a flat in acres.
///
/// Adding them into a single total is not a summary, it is a category error:
/// 27 acres of farmland and 400 sq yd of sites are not one number.
public enum HoldingKind: String, CaseIterable, Sendable {
    case farmland, plot, home, commercial

    public static func of(propertyType: String) -> HoldingKind {
        switch propertyType {
        case "open_plot": .plot
        case "commercial": .commercial
        case "flat", "independent_house", "villa": .home
        default: .plot
        }
    }

    public var label: String {
        switch self {
        case .farmland: "Agricultural land"
        case .plot: "Open plots & sites"
        case .home: "Houses & flats"
        case .commercial: "Commercial"
        }
    }

    /// What ONE of these is called. `label` names the category on a dashboard;
    /// this names the thing on its own screen.
    public var noun: String {
        switch self {
        case .farmland: "farmland"
        case .plot: "plot"
        case .home: "house"
        case .commercial: "commercial property"
        }
    }

    public var icon: String {
        switch self {
        case .farmland: "leaf.fill"
        case .plot: "square.dashed"
        case .home: "house.fill"
        case .commercial: "storefront.fill"
        }
    }

    /// The unit this kind is quoted in.
    public var unit: UnitKey {
        switch self {
        case .farmland: .acre
        case .plot: .sqyd
        case .home, .commercial: .sqft
        }
    }
}

/// The number that means something for a property, and what it is measured in.
///
/// For a building it is the BUILT-UP area — a flat's land share is meaningless
/// and usually zero. For a site it is the land. Falls back to whichever is
/// present, so a house with no built-up figure still shows its plot rather
/// than a blank.
public func headlineArea(propertyType: String,
                         landArea: Double, landUnit: String,
                         builtupArea: Double, builtupUnit: String) -> (value: Double, unit: UnitKey)? {
    let kind = HoldingKind.of(propertyType: propertyType)
    switch kind {
    case .home, .commercial:
        if builtupArea > 0 { return (builtupArea, unitKey(builtupUnit)) }
        if landArea > 0 { return (landArea, unitKey(landUnit)) }
        return nil
    default:
        if landArea > 0 { return (landArea, unitKey(landUnit)) }
        if builtupArea > 0 { return (builtupArea, unitKey(builtupUnit)) }
        return nil
    }
}

/// An extent written for a person: "418.5 Sq. yards", "30.00 Acres".
///
/// Acres carry two decimals because a tenth of an acre is 484 square yards and
/// people quote them that way; the smaller units are whole numbers, since half a
/// square foot is noise.
public func areaText(_ amount: Double, _ unit: UnitKey) -> String {
    // Trailing zeros are dropped, but a real fraction is NEVER rounded away:
    // a 418.5 sq. yd plot printed as "419" on the dashboard while its own
    // screen said 418.5 is the app disagreeing with itself about the size of
    // somebody's land.
    let digits: Int = {
        switch unit {
        case .acre, .hectare: return 2
        case .gunta, .cent: return 2
        default: return 1
        }
    }()
    let f = NumberFormatter()
    f.numberStyle = .decimal
    f.locale = Locale(identifier: "en_IN")
    f.minimumFractionDigits = 0
    f.maximumFractionDigits = digits
    let n = f.string(from: NSNumber(value: amount)) ?? String(amount)
    return "\(n) \(unit.label)"
}
